tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java

```diff
@@ -259,4 +263,36 @@ private void cleanUp() {
         return TransactionHookManager.getHooks();
     }
 
+    private Throwable getOriginalException(Throwable ex) {
+        if (ex instanceof TransactionException) {
+            return ex.getCause();
+        } else {
+            return ex;
+        }
+    }
+
+    private TransactionalExecutor.Code getRollbackCode(GlobalTransaction tx) throws TransactionException {
+        if (tx.getLocalStatus() == GlobalStatus.RollbackRetrying) {
+            if (tx.getRole() == GlobalTransactionRole.Launcher) {
+                return TransactionalExecutor.Code.RollbackRetrying;
+            } else {
+                return TransactionalExecutor.Code.ParticipantRollbackDone;
```

Comment:
This is the scenario where the rollback fails. You should not return to ParticipantRollbackDone.

Reply:
> This is the scenario where the rollback fails. You should not return to ParticipantRollbackDone.

changed